src/util/buffer.h

```diff
@@ -39,7 +40,7 @@ class buffer {
     void expand() {
         unsigned new_capacity = m_capacity << 1;
         T * new_buffer        = reinterpret_cast<T*>(memory::allocate(sizeof(T) * new_capacity));
-        memcpy(new_buffer, m_buffer, m_pos * sizeof(T));
+        std::uninitialized_copy(std::make_move_iterator(m_buffer), std::make_move_iterator(m_buffer + m_pos), new_buffer);
```

Comment:
@nunoplopes can you check this fix?

Reply:
@NikolajBjorner This change is analogue to what the LLVM people did in their [SmallVec class](http://llvm.org/doxygen/SmallVector_8h_source.html#l00194) when they encountered a similar issue.

Reply:
I have no doubts the fix is correct. Nuno changed this code last, so it is better to make sure he is aware.

Reply:
This change doesn't make sense. The memcpy is copying initialized memory only, so there's not undefined behavior here.

Reply:
There are two problems that are fixed by this change:
- Using `memcpy` to copy one initialized object into another is only legal if `T` is a [trivally copyable type](http://en.cppreference.com/w/cpp/types/is_trivially_copyable).
- `new_buffer` points to uninitialized memory, therefore new objects need to be created in the target buffer, which `memcpy` does not do.

Actually, looking at the code again, it might be more useful to either enforce [trivial destructible](http://en.cppreference.com/w/cpp/types/is_destructible) `T`s, to add destructor calls in `buffer<T>::free_memory`.

Reply:
I agree that memcpy is not always safe. However, not all code is Z3 is safe to change to something else, due to missing copy constructors, etc.
These kind of changes are actually high risk even though they don't look like. Without auditing all users of buffer, I can't sign off such a change. I suggest this change is dropped.

Reply:
fair enough, I am dropping the commit from the pull request